### PR TITLE
[SPARK-31595][SQL] Spark sql should allow unescaped quote mark in quoted string

### PR DESCRIPTION
**Pull request by adrian-wang**
### What changes were proposed in this pull request?
`def splitSemiColon` cannot handle unescaped quote mark like "'" or '"' correctly. When there are unmatched quotes in a string, `splitSemiColon` will not drop off semicolon as expected.

### Why are the changes needed?
Some regex expression will use quote mark in string. We should process semicolon correctly.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added Unit test and also manual test.

### CODE REVIEW

**Comment by adrian-wang** [comment]
@xuanyuanking

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 1]
```diff
@@ -519,13 +520,13 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
     for (index <- 0 until line.length) {
       if (line.charAt(index) == '\'' && !insideComment) {
         // take a look to see if it is escaped
```
Comment (dilipbiswal) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:522]: @adrian-wang Should we update the comment to reflect the newly added condition?

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 2]
```diff
@@ -519,13 +520,13 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
...
-        if (!escape) {
+        if (!escape && !insideDoubleQuote) {
           // flip the boolean variable
           insideSingleQuote = !insideSingleQuote
         }
       } else if (line.charAt(index) == '\"' && !insideComment) {
         // take a look to see if it is escaped
```
Comment (dilipbiswal) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:528]: @adrian-wang Same.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala (hunk 1 above)
Comment (xuanyuanking) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala:522]: Yep, maybe we can rephrase [this comment](https://github.com/apache/spark/pull/28393/files#diff-f7aac41bf732c1ba1edbac436d331a55R510) here.

**Review by xuanyuanking — COMMENTED** [review]
LGTM for the changes, cc @cloud-fan

**Comment by adrian-wang** [comment]
@dilipbiswal @xuanyuanking Thanks for your advice, I have updated the code.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 3]
```diff
@@ -500,4 +500,13 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
         |   ;""".stripMargin -> "testcomment"
     )
   }
+
+  test("Should allow unescaped quote mark in quoted string") {
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:504]: Plz add a prefix `SPARK-31595: Should allow...`.

**Comment by dilipbiswal** [comment]
LGTM
cc @maropu

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 4]
```diff
@@ -500,4 +500,13 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
         |   ;""".stripMargin -> "testcomment"
     )
   }
+
+  test("SPARK-31595 Should allow unescaped quote mark in quoted string") {
+    runCliWithin(1.minute)(
+      """SELECT '"legal string a';select 1 + 234;""".stripMargin -> "235"
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:506]: Is this an only issue in the thrift server side? How about the spark side?
```
scala> sql("""SELECT '"legal string a';select 1 + 234;""").show()
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input 'select' expecting {<EOF>, ';'}(line 1, pos 25)

== SQL ==
SELECT '"legal string a';select 1 + 234;
-------------------------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:268)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:135)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:49)
```

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 5]
```diff
@@ -500,4 +500,13 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
         |   ;""".stripMargin -> "testcomment"
     )
   }
+
+  test("SPARK-31595 Should allow unescaped quote mark in quoted string") {
+    runCliWithin(1.minute)(
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:505]: Could you explicitly set false at `spark.sql.parser.escapedStringLiterals` for the tests below?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 4 above)
Comment (adrian-wang) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:506]: This is not the same case. `sql()` only accepts single sql statement, even
```scala
sql("select 1; select 2;")
```
will return error.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 5 above)
Comment (adrian-wang) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:505]: Actually this has nothing to do with `spark.sql.parser.escapedStringLiterals`, even if the config is set to true, the parser should accept this string.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 4 above)
Comment (maropu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:506]: Ah, I see.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 5 above)
Comment (maropu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:505]: Ur, I got it. I think the option is misleading... Could you remove it in the PR descritpion?

Comment (adrian-wang) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:505]: Done that, thanks!

**Review by maropu — APPROVED** [review]
Looks fine and thanks for the update, @adrian-wang cc: @wangyum

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 6]
```diff
@@ -500,4 +500,13 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
...
+      """SELECT '"legal string a';select 1 + 234;""".stripMargin -> "235"
+    )
+    runCliWithin(1.minute)(
+      """SELECT "legal 'string b";select 22222 + 1;""".stripMargin -> "22223"
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:509]: nit: let's not use the multiline string style for a single line string.

Comment (adrian-wang) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Updated, thanks!

**Comment by cloud-fan** [comment]
cc @juliuszsompolski as well

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**